### ISSUE TITLE
Forcibly republish the package with the same version

### ISSUE THREAD

**Issue by egel** [verdaccio/verdaccio#1402]
**Is your feature request related to a problem? Please describe.**
Let's have a situation that I want to keep my package version in sync with other software. I.e.: my-dto:1.2.3, other-software:1.2.3. The problem that I have is once I publish the package I can't easily unpublish it nor forcibly update it. The question is: Is there any way to forcibly republish the package in my private verdaccio instance with the same version?

**Describe the solution you'd like**
I want to be able to forcibly update the package which has already been published.

**Describe alternatives you've considered**
or at least be able to unpublish it quickly and then publish it again, so I don't need to remove entries from the database to erase old package.

**Additional context**

- I stripped fragile data with XXX.
- My Verdaccio version: 3.11.4

1. The package is already published by in my registry under `https://npm.XXX.systems/@my-dto`

```
yarn publish --registry=https://npm.XXX.systems
yarn publish v1.17.3
[1/4] Bumping version...
info Current version: 9.4.0
question New version: 9.4.0
[2/4] Logging in...
[3/4] Publishing...
error Couldn't publish package: "https://npm.XXX.systems/@my%2fdto: this package is already present"
info Visit https://yarnpkg.com/en/docs/cli/publish for documentation about this command.
```

**Comment by kopax** [comment]
I think this is good to not republish a package. Use test to avoid publishing a package with a wrong version. If you fix, then publish a new version. `npm` allow to unpublish and republish package in a short amount of time. This is because package maintainer can send hack in their republish version and other people won't notice.

**Comment by Qix-** [comment]
Please remember that people use Verdaccio for development. Trying to add a feature to @storybookjs is currently a **massive** headache - I have spent the last two hours trying to figure out how to get Verdaccio to play nice.

I'm developing locally. I don't care about npm's silly package publishing policies. If I write `--force`, **I mean it**. Let me publish over versions without having to unpublish.

If you're really scared about this, at least let me specify it in the Verdaccio configuration that force publishes are OK.

**Comment by kopax** [comment]
what is wrong with `npm unpublish` ?

**Comment by juanpicado** [comment]
> what is wrong with npm unpublish ?

`npm unpublish` would be a 2x step process, which some environments cannot afford, for instance CI or just annoying increase the version manually if you don't use Lerna or something sophisticated .

Also there is a trade off with such command, I refer to the official doc.

>If no version is specified, or if all versions are removed then the root package entry is removed from the registry entirely.

Thus, Verdaccio will remove the whole package if you had only one version. Thus, `npm publish --force` becomes handy.

So, referring to then @egel pain, I had also that pain and we have been discussing internally how to achieve it in a safe way and by demand. One of the ideas is introduce the `development` mode, either via configuration or flag, eg:

```
$> verdaccio --dev
```

This is still just an idea and has a lot of tradeoff, for instance uplinks, possibles side effects we need to research first.

**Comment by DanielRuf** [comment]
A dev mode sounds good.

**Comment by egel** [comment]
Is there any progress with the issue?Writing due to bot reminder.

**Comment by juanpicado** [comment]
@egel it's on the queue, but no progress yet. Sorry for the bot, should not be trigger if is already scheduled for dev.

**Comment by haquezameer** [comment]
+1 to @Qix- 

We use `verdaccio` for E2E tests. If add a `uplink` to the package, it won't allow me to publish even once into the locally running `verdaccio` instance, even though it does not contain **any** versions of the package. Because of the `uplink` it hits the uplink and downloads the version of the package and returns error saying `can't publish over existing version`. I would expect verdaccio to not hit the uplink in which case it would allow me to publish.

**Comment by kachkaev** [comment]
@haquezameer same issue here. Have you found a workaround?

**Comment by haquezameer** [comment]
@kachkaev Yes. So I actually separated my verdaccio configs into two separate configs: `verdaccio-publish.yaml` and `verdaccio-install.yaml`.

I only add the uplinks config in `verdaccio-install.yaml` and do not add any uplinks in `verdaccio-publish.yaml`. This prevents verdaccio from connecting to the uplink when publishing and allows me to publish to the local registry. Then while installing I start verdaccio with `verdaccio-install.yaml`.

**Comment by jampy** [comment]
I'm interested in the same mechanism.

I'm wondering: when republishing with the same version, wouln't this cause a lockfile checksum violation when re-installing the package? I don't know exactly what npm/yarn/etc.. do when the checksum does not match, but I expect them to abort the installation.

**Comment by eldoy** [comment]
`npm unpublish --force` and then republish seems to work?

**Comment by itineric** [comment]
Also interested in this feature.
It probably should be a "per-package" configuration?!
But I do not understand why this should be associated with a "dev" mode?

As mentioned before, CI may have to republish in some cases.

Other registry implementations allow this (on option) and npm install does not fail because of checksum change for the same package.
Im voting for this optionnaly available out-of-the-box.

**Comment by monolithed** [comment]
Kind of weird to explain why `--force` is needed. It's a standard flag, it works in Nexus too. 
What problem does this limitation actually solve?

**Comment by randomhash** [comment]
It solves supply chain attacks to some extent

**Comment by DanielRuf** [comment]
@randomhash are you referring to the current cases with axios and others?
Verdaccio is a registry that you can self-host in your company.

If there is someone force-publishing a package to your Verdaccio instance, you might have bigger problems than just this.
Because then they have access to much more in most cases.

Verdaccio is not used by npmjs, so that would not prevent any of the latest supply chain attacks in my opinion.

Or to focus on npmjs: you can not republish the same package version there in general. So why would you want to forcibly republish the same version internally, even if it would have been deleted on npmjs? I do not really see, how Verdaccio could have helped here, especially when you mean the missing `--force` feature. Might be also less secure, when `--force` is possible creating new possible attacks and impacts your threat model.

**Comment by DanielRuf** [comment]
And so far Verdaccio is not widely used in many companies. We still had other enterprise solutions there. One of them was mentioned here: https://github.com/verdaccio/verdaccio/issues/1402#issuecomment-3395563461

Please keep in mind, that Verdaccio is an OSS project created and maintained by people working on it in their personal free time. And the requirements are different compared to these enterprise solutions - might also apply to experience and personal needs.

**Comment by mbtools** [comment]
If you use Verdaccio as your registry (with uplink to npmjs), you can use the package filter plugin (planned for v7, https://github.com/verdaccio/verdaccio/issues/5680) to delay the availability of packages. A delay of a few days would allow the most prominent supply chain attacks to be mitigated before they would hit your machine.

**Comment by DanielRuf** [comment]
Ah, good hint. Newer versions of different package managers also have some `minimumReleaseAge` setting. Could be combined with this.

**Comment by juanpicado** [comment]
V6 also will have it as optional, just I haven't released it yet.
